Public/Src/Utilities/Native/IO/Helpers.cs

```diff
@@ -135,12 +138,70 @@ public static class Helpers
                     // Only log exceptions if they are not being re-thrown already
                     if (logExceptions)
                     {
-                        Tracing.Logger.Log.RetryOnFailureException(Events.StaticContext, e.Message ?? e.ToString());
+                        Tracing.Logger.Log.RetryOnFailureException(Events.StaticContext, e.ToString());
                     }
                 }
             }
 
             return done;
         }
+
+        /// <summary>
+        /// A generic version of <see cref="RetryOnFailure"/>.
+        /// </summary>
+        public static Possible<TResult, Failure> RetryOnFailure<TResult>(
+          Func<bool, Possible<TResult, Failure>> work,
+          bool logExceptions = true,
+          bool rethrowExceptions = false,
+          int numberOfAttempts = DefaultNumberOfAttempts,
+          int initialTimeoutMs = DefaultInitialTimeoutMs,
+          int postTimeoutMultiplier = DefaultPostTimeoutMultiplier)
+        {
+            Possible<TResult, Failure>? possiblyResult = null;
+
+            bool success = RetryOnFailure(
+                isLastRetry =>
+                {
+                    possiblyResult = work(isLastRetry);
+                    return possiblyResult.Value.Succeeded;
+                },
+                logExceptions, rethrowExceptions, numberOfAttempts, initialTimeoutMs, postTimeoutMultiplier);
+
+            if (!possiblyResult.HasValue)
+            {
+                return new Failure<string>($"Exhausted all of the '{numberOfAttempts}' retry attempts.{(logExceptions ? " Look at the log for more details." : string.Empty)}");

```

Comment:
I think if possiblyResult is null then there is a bug in the code (like the number of attempts is 0.

I think we can enforce this by using `Contract.Assert(possiblyResult != null)`.

Reply:
Discussed offline.

---
In reply to: [273276304](https://github.com/Microsoft/BuildXL/pull/104#discussion_r273276304) [](ancestors = 273276304)